csi/controller_server.go

```diff
@@ -647,6 +690,68 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 	return rsp, nil
 }
 
+func (cs *ControllerServer) autoAttachVolumeToRandomNode(volume *longhornclient.Volume) (nodeID string, err error) {
+	nodeCollection, err := cs.apiClient.Node.List(longhornclient.NewListOpts())
+	if err != nil {
+		return "", err
+	}
+	engineImage, err := cs.apiClient.EngineImage.ById(types.GetEngineImageChecksumName(volume.CurrentImage))
+	if err != nil {
+		return "", err
+	}
+	nodeID, err = findARandomReadyNode(engineImage, nodeCollection)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get random node to attach to volume %v", volume.Name)
+	}
+	_, err = cs.publishVolume(volume, nodeID, func() error {
+		checkVolumePublished := func(vol *longhornclient.Volume) bool {
+			return isVolumeAvailableOn(vol, nodeID) || isVolumeShareAvailable(vol)
+		}
+		if !cs.waitForVolumeState(volume.Name, "volume published", checkVolumePublished, false, false) {
+			return status.Errorf(codes.DeadlineExceeded, "volume %s failed to attach to node %s", volume.Name, nodeID)
+		}
+		return nil
+	})
+	if err != nil {
+		return "", err
+	}
+	return nodeID, nil
+}
+
+// findARandomReadyNode is similar to FindARandomReadyNode but with
+// longhornclient since not possible without introducing longhorn manager client.
+func findARandomReadyNode(ei *longhornclient.EngineImage, nc *longhornclient.NodeCollection) (string, error) {
```

Comment:
I still have a concern about a duplicated code even in CSI node plugin we only use longhorn client only.

Is it possible to extend longhorn manager `VolumeAttach` capability to support attaching to a random node if node id is empty? then you can simply call the volume attach method of longhorn client to request to attach to a random node instead of duplicating same code w/ different data structure here.

Reply:
I don't think it'sa good idea. Actually I already discussed this part with Chin-ya. Ideally it's Longhorn backend's responsibility to do auto attachment. Asking the CSI plugin (for CSI snapshot) or recurring pod (for backup recurring jobs) to send attach/detach calls is a short-term solution. We can re-visit it after refactoring the volume controller.

Reply:
Sure, also try to avoid any duplicated code existing when revisiting this part later.

Reply:
Removed this function instead use the one in datastore. The recurring job can leverage lhClient to get the nodes and engine image.